src/adapter/src/coord/appends.rs

```diff
@@ -248,6 +245,30 @@ impl<S: Append + 'static> Coordinator<S> {
             return;
         }
 
+        let _write_lock_guard = if let Some(guard) = self
+            .pending_writes
+            .iter_mut()
+            .filter_map(|write| write.take_write_lock())
+            .next()
+        {
+            // If some pending transaction already holds the write lock, then we can execute a group
+            // commit.
+            guard
+        } else if let Ok(guard) = Arc::clone(&self.write_lock).try_lock_owned() {
+            // If no pending transaction holds the write lock, then we need to acquire it.
+            guard
+        } else {
+            // If some running transaction already holds the write lock, then one of the
+            // following things will happen:
+            //   1. The transaction will submit a write which will transfer the
+            //      ownership of the lock to group commit and trigger another group
+            //      group commit.
+            //   2. The transaction will complete without submitting a write (abort,
+            //      empty writes, etc) which will drop the lock. The deferred group
+            //      commit will then acquire the lock and execute a group commit.
+            return self.defer_write(Deferred::GroupCommit);
```

Comment:
This will not work for DDL I've just realized. If we can't get the lock then we should continue with the DDLL and not process writes to user tables.

Reply:
Fixed.